vcenter_operator/configurator.py

```diff
@@ -58,45 +58,72 @@ def __init__(self, domain, global_options={}):
         self.global_options['cells'] = {}
         self.global_options['domain'] = domain
 
-    def __call__(self, added, removed):
-        for name in added:
-            try:
-                host = '{}.{}'.format(name, self.domain)
-                if host in self.vcenters:
-                    continue
-
-                # Vcenter doesn't accept / in password
-                password = self.mpw.derive('long', host).replace("/", "")
+        atexit.register(self._disconnect_vcenters)
 
-                LOG.info("{}".format(host))
-                if hasattr(ssl, '_create_unverified_context'):
-                    context = ssl._create_unverified_context()
-
-                    service_instance = SmartConnect(host=host,
-                                                    user=self.username,
-                                                    pwd=password,
-                                                    port=443,
-                                                    sslContext=context)
-
-                if service_instance:
-                    atexit.register(Disconnect, service_instance)
-
-                    self.vcenters[host] = {
-                        'service_instance': service_instance,
-                        'username': self.username,
-                        'password': password,
-                        'host': host,
-                        'name': name,
-                    }
+    def _disconnect_vcenters(self):
+        """Disconnect all vcenters we are connected to"""
+        for host in six.iterkeys(self.vcenters):
+            service_instance = self.vcenters[host]['service_instance']
+            if not service_instance:
+                continue
+            try:
+                Disconnect(service_instance)
+            except:
+                # best effort disconnection
+                pass
 
-            except vim.fault.InvalidLogin as e:
-                LOG.error("%s: %s", host, e.msg)
-            except (Exception, socket_error) as e:
-                LOG.error("%s: %s", host, e)
+    def __call__(self, added, removed):
+        """Add/remove vcenters from our managed list of vcenters"""
+        for name in added:
+            host = '{}.{}'.format(name, self.domain)
+            self._reconnect_vcenter_if_necessary(host)
 
         if removed:
             LOG.info("Gone vcs {}".format(removed))
 
+    def _connect_vcenter(self, host):
+        """Create a connection to host and add it to self.vcenters"""
+        try:
+            # Vcenter doesn't accept / in password
+            password = self.mpw.derive('long', host).replace("/", "")
+
+            LOG.info("Connecting to {}".format(host))
+            if hasattr(ssl, '_create_unverified_context'):
```

Comment:
If ssl does not have this attribute then `service_instance` will never be set and the check in line 100 will break. I'd recommend either removing this check or setting `service_instance` to None in an else clause.

Reply:
I've set it to `None` as default. Thank you.